### PR TITLE
8305083: Remove finalize() from test/hotspot/jtreg/vmTestbase/nsk/share/ and /jpda that are used in serviceability/dcmd/framework tests

### PR DESCRIPTION
**Pull request by afshin-zafari**
The `finalize()` method is removed from base classes/interfaces and are replaced by a Cleaner callback..

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305083](https://bugs.openjdk.org/browse/JDK-8305083): Remove finalize() from test/hotspot/jtreg/vmTestbase/nsk/share/ and /jpda that are used in serviceability/dcmd/framework tests

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) ⚠️ Review applies to [fa8e4537](https://git.openjdk.org/jdk/pull/13420/files/fa8e453767a38d6e72f3a8b2065a6659b68466c6)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**) ⚠️ Review applies to [fa8e4537](https://git.openjdk.org/jdk/pull/13420/files/fa8e453767a38d6e72f3a8b2065a6659b68466c6)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13420/head:pull/13420` \
`$ git checkout pull/13420`

Update a local copy of the PR: \
`$ git checkout pull/13420` \
`$ git pull https://git.openjdk.org/jdk.git pull/13420/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13420`

View PR using the GUI difftool: \
`$ git pr show -t 13420`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13420.diff">https://git.openjdk.org/jdk/pull/13420.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13420#issuecomment-1504941466)

### CODE REVIEW

**Comment by plummercj** [comment]
The interaction of Finalizable, FinalizableObject, Finalizer, and FinalizerThread is hard to follow, so it is hard to tell if your changes still honor the needed "finalization". Did you figure this all out? Perhaps you can summarize how it worked properly when using finalize(), and why it continues to work after your changes. I'm especially suspicious of your removal of the finalize() call in FinalizableObject. I don't see how this is made up for elsewhere in your changes. Is it because finalizeAtExit() is always overridden to do the right thing? If so, perhaps it should be made abstract in FinalizableObject.

**Comment by dholmes-ora** [comment]
> The interaction of Finalizable, FinalizableObject, Finalizer, and FinalizerThread is hard to follow,

These classes implement a cleanup-at-vm-exit mechanism, which uses `finalize` as the cleanup method. It is unclear if the classes that implement `FinalizableObject` can actually get cleaned up before VM exit through normal GC and execution of finalizers. Simply adding a Cleaner does not provide the cleanup-at-vm-exit functionality. You need to rename Finaliz* to Clean* and keep all these classes, as well as adding a Cleaner - which should be done as part of CleanableObject.

**Comment by coleenp** [comment]
Another change that has turned into something much bigger.  This cleanup is worth doing.  Possible steps are (not in order and not in one push):

1.  It seems like FinalizerObject should be removed in favor of classes that implement interface Finalizable and are forced to implement finalizeAtExit, so the Finalize thread can call the function on all the objects that get registered to it.  Not sure if this mechanism is useful or not.  Maybe it should be removed.  If the process exits, are these connections closed?
2. If this mechanism is needed, rename Finalizer, FinalizerThread and Finalizable (ugh) to TestCleaner{Thread}. Cleanable with cleanAtExit, and rewrite/remove all references to these in the comments.   Or just delete this (better?)
3. Fix remaining comments and especially all the typos.
4. This patch removes the classes that have registered finalize methods in favor of the cleaner mechanism, which looks good to me.  These appear to be cases where some connection needs to be closed, or something like that.

It seems to me that step 4 can be done as in this patch, leaving 1, 2 and 3 to further cleanups and RFEs. Looks like we've uncovered a mess.  The purpose of these changes are to remove the dependence on finalizer() with their associated warnings.

**Comment by coleenp** [comment]
Amendment to my comment above.  1. FinalizableObject does seem to serve a purpose, so shouldn't be removed.

File: test/hotspot/jtreg/vmTestbase/nsk/share/FinalizableObject.java [hunk 1]
```diff
@@ -22,29 +22,47 @@
...
+ * <code>Finalizable</code>.
  *
  * @see Finalizable
  * @see Finalizer
  */
 public class FinalizableObject implements Finalizable {
 
+
+    /**
+     * This method will be invoked by <tt>Finalizer</tt> when virtual mashine
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/FinalizableObject.java:39]: You copied the `mashine` typo here.

File: test/hotspot/jtreg/vmTestbase/nsk/share/FinalizableObject.java [hunk 2]
```diff
@@ -22,29 +22,47 @@
...
+     * shuts down.
+     *
+     * @throws Throwable if any throwable exception thrown during finalization
+     *
+     * @see Finalizer
+     */
+
+    public void cleanup() {}
     /**
      * This method will be invoked by <tt>Finalizer</tt> when virtual mashine
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/FinalizableObject.java:49]: Another `mashine` typo.

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java [hunk 3]
```diff
@@ -33,11 +33,18 @@
...
      * @throws Throwable if any throwable exception thrown during finalization
      */
     public void finalizeAtExit() throws Throwable;
 
+    /*
+     * This method will register a cleanup method and creates an instance of Finalizer
+     * to register the object for finalization at VM exit.
+     * It is implemented in FinalizableObject.
+     */
+    public void registerCleanup();
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:48]: Can you not implement this as a default method?

**Review by dholmes-ora — COMMENTED** [review]
Not sure about keeping the "finalize" terminology - though perhaps with more extensive commenting in the key classes it may be okay.

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java (hunk 3 above)
Comment (coleenp) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:48]: It seems much better to force implementors to provide this method.

**Review by coleenp — APPROVED** [review]
This looks good to me with correcting mashine typo.  We should file another RFE to rename these classes and methods and fix all the comments, many containing typos and let the hotspot/srv take care of that.

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java (hunk 3 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:48]: Why? They all seem to do exactly the same thing.

Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:48]: After I moved the `registerCleanup` to the body of a `default` method in the interface, there is no need for the implementors of the `Finalizable` interface to provide this method. All of them can use the default one.

Comment (dholmes-ora) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:48]: > All of them can use the default one

Exactly my point.

Comment (coleenp) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:48]: Default methods for interface classes were invented to solve a problem of compatibility if I remember correctly.  Forcing subclasses to implement the interface method or have a superclass of the subclass to implement the interface method seems like it avoids the problem of silently not registering the cleanup or action that the interface method should force you to do.  To solve the duplicated registerCleanup() cases, the two other classes could extend FinalizableObject then inherit its implementation of registerCleanup().

**Review by coleenp — APPROVED** [review]

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java [hunk 4]
```diff
@@ -33,11 +33,18 @@
...
-     * This method will be invoked by <tt>Finalizer</tt> when virtual mashine
+     * This method will be invoked by <tt>Finalizer</tt> when virtual machine
      * shuts down.
      *
      * @throws Throwable if any throwable exception thrown during finalization
      */
     public void finalizeAtExit() throws Throwable;
 
+    /*
+     * This method will register a cleanup method and creates an instance of Finalizer
```
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:44]: Nit: `creates` => `create`

File: test/hotspot/jtreg/vmTestbase/nsk/share/LocalProcess.java [hunk 5]
```diff
@@ -166,12 +164,10 @@ protected void kill() {
     }
 
     /**
-     * Finalize mirror by invoking <code>close()</code>.
+     * This method is called at finalization and calls  <code>kill()</code>.
```
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/share/LocalProcess.java:167]: Nit: Extra space after `and calls`.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/BindServer.java [hunk 6]
```diff
@@ -95,6 +96,7 @@ public class BindServer implements Finalizable {
     private int unauthorizedRequests = 0;
     private int busyRequests = 0;
 
+
```
Comment (sspitsyn) [inline test/hotspot/jtreg/vmTestbase/nsk/share/jpda/BindServer.java:99]: Nit: Unneeded extra line.

**Review by sspitsyn — APPROVED** [review]
Looks okay to me but posted some nits.
Thanks,
Serguei

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java (hunk 3 above)
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:48]: > To solve the duplicated registerCleanup() cases, the two other classes could extend FinalizableObject then inherit its implementation of registerCleanup().

The other two cases already extend the `Log.Logger` and cannot extend `FinalizableObject`.

**Comment by dholmes-ora** [comment]
> Default methods for interface classes were invented to solve a problem of compatibility if I remember correctly. 

Yes they were added a way to extend existing interfaces, but the point is that a default method provides an implementation that will work "good enough" for any implementing class in general - and is this case the default implementation is all that is needed (it is like adding a method to a common base class).

**Comment by coleenp** [comment]
You're going to end up with things not overriding this method that should.

**Comment by dholmes-ora** [comment]
> You're going to end up with things not overriding this method that should.

??? We are controlling all the classes - we know if anything would need to have a different implementation of this method.

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java (hunk 3 above)
Comment (coleenp) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:48]: I though default methods were bad design just like default parameters.  maybe not.  I didn’t think yours would compile either but I guess it does.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/BindServer.java (hunk 6 above)
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jpda/BindServer.java]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/share/LocalProcess.java (hunk 5 above)
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/LocalProcess.java]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java (hunk 4 above)
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java [hunk 7]
```diff
@@ -22,29 +22,47 @@
...
+    default public void finalizeAtExit() throws Throwable {
+        cleanup();
+    }
 
-    /*
-     * This method will register a cleanup method and creates an instance of Finalizer
+    /**
+     * This method will register a cleanup method and create an instance of Finalizer
      * to register the object for finalization at VM exit.
      * It is implemented in FinalizableObject.
```
Comment (coleenp) [inline test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:56]: This sentence can go since it's implemented here now.

File: test/hotspot/jtreg/vmTestbase/nsk/share/MainWrapper.java [hunk 8]
```diff
@@ -43,11 +43,10 @@ public static void main(String[] args) throws Throwable {
 
         // It is needed to register finalizer thread in default thread group
         // So FinalizerThread thread can't be in virtual threads group
-        Finalizer finalizer = new Finalizer(new FinalizableObject());
-        finalizer.activate();
 
-        // Some tests use this property to understand if virtual threads are used
-        System.setProperty("main.wrapper", wrapperName);
```
Comment (coleenp) [inline test/hotspot/jtreg/vmTestbase/nsk/share/MainWrapper.java:50]: Should this line have been deleted?

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java [hunk 9]
```diff
@@ -84,6 +85,10 @@ abstract public class DebugeeProcess extends FinalizableObject {
     protected DebugeeProcess (DebugeeBinder binder) {
         this.binder = binder;
         this.log = binder.getLog();
+
+        // As the alternative to finalize(), register the cleanup() method
+        // to be called when this instance becomes unreachable.
+        Cleaner.create().register(this, () -> cleanup());
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java:91]: Why do we need to do this explicitly here? Why not call `registerCleaner`?

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/SocketIOPipe.java [hunk 10]
```diff
@@ -23,6 +23,7 @@
 package nsk.share.jpda;
 
 import java.io.IOException;
+import java.lang.ref.Cleaner;
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/vmTestbase/nsk/share/jpda/SocketIOPipe.java:26]: Not needed.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
This is a tricky issue to solve cleanly and minimally. If every class extended FinalizableObject then all the implementation could go there, but we have to try and split things across FinalizableObject and Finalizable because some classes only implement the interface - this leads to some unfortunate design choices.

I think a better design for the classes that can't extend FinalizableObject would be for them to contain a FinalizableObject, the cleanup action for which would cleanup the host object. That could allow the removal of the Finalizable interface and simplify the general usage patterns. But that may be going too far for this particular PR ... ?

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java [hunk 11]
```diff
@@ -22,29 +22,47 @@
  */
 
 package nsk.share;
+import java.lang.ref.Cleaner;
 
 /**
  * Finalizable interface allows <tt>Finalizer</tt> to perform finalization of an object.
  * Each object that requires finalization at VM shutdown time should implement this
- * interface and activate a <tt>Finalizer</tt> hook.
+ * interface and call the <tt>registerClenup</tt> to activate a <tt>Finalizer</tt> hook.
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:30]: Corrected.

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java [hunk 12]
```diff
@@ -22,29 +22,47 @@
...
+    /**
+     * This method will register a cleanup method and create an instance of Finalizer
      * to register the object for finalization at VM exit.
      * It is implemented in FinalizableObject.
+     *
+     * @see Finalizer
      */
-    public void registerCleanup();
+    default public void registerCleanup() {
+       // install finalizer to print errors summary at exit
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:60]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java [hunk 13]
```diff
@@ -22,29 +22,47 @@
...
+       Finalizer finalizer = new Finalizer(this);
+       finalizer.activate();
 
+       // register the cleanup method to be called when this Log instance becomes unreachable.
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java:64]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/share/FinalizableObject.java [hunk 14]
```diff
@@ -33,36 +32,11 @@
...
-
     /**
-     * This method will be invoked by <tt>Finalizer</tt> when virtual mashine
-     * shuts down.
-     *
-     * @throws Throwable if any throwable exception thrown during finalization
-     *
-     * @see Finalizer
+     * All instances of this class, should implement their own cleanup method
+     * to clean appropriately the objects they used.
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/FinalizableObject.java:37]: Suggested text is replaced.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/BindServer.java [hunk 15]
```diff
@@ -406,21 +406,22 @@ public synchronized void close() {
      * Make finalization of <code>BindServer</code> object by invoking
      * method <code>close()</code>.
      *
+     * This is replacement of the deprecated finalize() and is called
+     * when this instance becomes unreachable.
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jpda/BindServer.java:410]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeBinder.java [hunk 16]
```diff
@@ -551,20 +551,12 @@ public void close() {
...
-     */
-    protected void finalize() throws Throwable {
-        close();
-        super.finalize();
-    }
-
-    /**
-     * Finalize binder at exit by invoking <code>finalize()</code>.
+     * This is replacement of the finalize() method and is called when this
+     * instance becomes unreachable.
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeBinder.java:561]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java [hunk 17]
```diff
@@ -84,6 +85,10 @@ abstract public class DebugeeProcess extends FinalizableObject {
     protected DebugeeProcess (DebugeeBinder binder) {
         this.binder = binder;
         this.log = binder.getLog();
+
+        // As the alternative to finalize(), register the cleanup() method
```
[the remark this answers is not in the document]
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java:89]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/SocketIOPipe.java (hunk 10 above)
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jpda/SocketIOPipe.java:26]: Done.

File: test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java (hunk 9 above)
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/jpda/DebugeeProcess.java:91]: Was remained from the former solution. `registerCleanup()` is called.

File: test/hotspot/jtreg/vmTestbase/nsk/share/MainWrapper.java (hunk 8 above)
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/MainWrapper.java:50]: Came after merging with master.

File: test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java (hunk 7 above)
Comment (afshin-zafari) [reply test/hotspot/jtreg/vmTestbase/nsk/share/Finalizable.java]: Done.

**Comment by afshin-zafari** [comment]
A new PR: https://github.com/openjdk/jdk/pull/13884, is created for this issue. 
This branch is far behind the master.

**Comment by afshin-zafari** [comment]
> This is a tricky issue to solve cleanly and minimally. If every class extended FinalizableObject then all the implementation could go there, but we have to try and split things across FinalizableObject and Finalizable because some classes only implement the interface - this leads to some unfortunate design choices.
> 
> I think a better design for the classes that can't extend FinalizableObject would be for them to contain a FinalizableObject, the cleanup action for which would cleanup the host object. That could allow the removal of the Finalizable interface and simplify the general usage patterns. But that may be going too far for this particular PR ... ?

Please, continue this discussion on the new PR: https://github.com/openjdk/jdk/pull/13884. Thanks!
